### PR TITLE
SPIN-1387 parallel stage execution

### PR DESCRIPTION
**Pull request by robfletcher**

### CODE REVIEW

**Comment by robfletcher** [comment]
Updated to make synthetic stages work properly. PTAL @ajordens

File: orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/ExecutionRunnerSupport.java [hunk 1]
```diff
@@ -36,13 +34,27 @@ public ExecutionRunnerSupport(Collection<StageDefinitionBuilder> stageDefinition
...
-  public <T extends Execution> void start(T execution) throws Exception {
+  /**
+   * The default implementation plans all stages.
+   * Implementations may override this method completely or use it to plan the entire execution before running it.
+   *
+   * @param execution the execution instance with skeleton stages.
+   * @param <T>       the execution type.
+   * @throws Exception
+   */
+  @Override public <T extends Execution<T>> void start(T execution) throws Exception {
```
Comment (ajordens) [inline orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/ExecutionRunnerSupport.java]: Was thinking (the other day) if `start` is the appropriate name for this method vs. `plan`?

It's overridden in subclasses (right?) with a super.start(execution) and then the behavior for actually running it in spring batch.

File: orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/model/AbstractStage.groovy [hunk 2]
```diff
@@ -27,42 +30,34 @@ import com.netflix.spinnaker.orca.ExecutionStatus
...
   String name
   Long startTime
   Long endTime
   ExecutionStatus status = NOT_STARTED
-  void setStatus(ExecutionStatus status) {
-    this.status = status
-  }
-  @JsonBackReference
-  T execution
+  @JsonBackReference T execution
```
Comment (ajordens) [inline orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/model/AbstractStage.groovy]: Never been a big fan of the annotations on same line as declaration ...

File: orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/model/AbstractStage.groovy [hunk 3]
```diff
@@ -27,42 +30,34 @@ import com.netflix.spinnaker.orca.ExecutionStatus
...
+  Collection<String> requisiteStageRefIds = []
+  SyntheticStageOwner syntheticStageOwner
   List<InjectedStageConfiguration> beforeStages = []
   List<InjectedStageConfiguration> afterStages = []
   long scheduledTime
 
   @JsonIgnore
   AtomicInteger stageCounter = new AtomicInteger(0)
 
+  @Deprecated
```
Comment (ajordens) [inline orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/model/AbstractStage.groovy]: Do we not need this anymore?

File: orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/model/AbstractStage.groovy [hunk 4]
```diff
@@ -98,15 +93,7 @@ abstract class AbstractStage<T extends Execution> implements Stage<T>, Serializa
   }
 
   AbstractStage(Execution execution, String type) {
-    this(execution, type, EMPTY_MAP)
-  }
-
-  @Override
-  Stage preceding(String type) {
```
Comment (ajordens) [inline orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/model/AbstractStage.groovy]: This `preceding` is a horrible method ... only works consistently if the pipeline isn't parallel (which all should).

File: orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/model/Stage.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+  /**
+   * sets the execution status for this stage
+   */
+  void setStatus(ExecutionStatus status);
+
+  /**
+   * Gets the last stage preceding this stage that has the specified type.
+   */
+  default Stage preceding(String type) {
+    int i = getExecution()
```
Comment (ajordens) [inline orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/model/Stage.java]: This is still a poor method.

If anything delegating to `ancestors()` seems more appropriate (provided ancestors handles the situation where there are no refIds on stages).

File: orca-spring-batch/src/main/java/com/netflix/spinnaker/orca/batch/SpringBatchExecutionRunner.java [hunk 6]
```diff
@@ -127,23 +165,124 @@ private Step initStep() {
       .build();
   }
 
-  private <E extends Execution> FlowBuilder<FlowJobBuilder> buildStepsForStage(FlowBuilder<FlowJobBuilder> flow, Stage<E> stage) {
+  private <E extends Execution<E>, Q> FlowBuilder<Q> buildStepsForStageAndDownstream(FlowBuilder<Q> flow, Stage<E> stage, Set<Serializable> alreadyBuilt) {
+    if (alreadyBuilt.contains(stage.getRefId())) {
+      log.info("Already built {}", stage.getRefId());
```
Comment (ajordens) [inline orca-spring-batch/src/main/java/com/netflix/spinnaker/orca/batch/SpringBatchExecutionRunner.java]: DEBUG maybe ... going to get verbose.

**Comment by ajordens** [comment]
LGTM but definitely need to run through a significant variety of pipelines both `linear` and `parallel` to ensure behavior hasn't changed.

Hopefully the generic/stage work won't increase the difficulty in rebasing off master until this is in a position to get merged down.

**Comment by ajordens** [comment]
The direction that this is headed in (_and the general cleanup we've been making along the way_) is great and much needed in `orca`.

File: orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/model/AbstractStage.groovy (hunk 2 above)
Comment (robfletcher) [inline orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/model/AbstractStage.groovy]: Probably just my auto-format in IntelliJ

File: orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/ExecutionRunnerSupport.java (hunk 1 above)
Comment (robfletcher) [inline orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/ExecutionRunnerSupport.java]: Correct.

File: orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/model/AbstractStage.groovy (hunk 3 above)
Comment (robfletcher) [inline orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/model/AbstractStage.groovy]: It's only used in `StageBuilder` so we can remove it once that's redundant.

File: orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/model/Stage.java (hunk 5 above)
Comment (robfletcher) [inline orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/model/Stage.java]: Yeah, I don't disagree but it's not a change in behavior here.

File: orca-spring-batch/src/main/java/com/netflix/spinnaker/orca/batch/SpringBatchExecutionRunner.java (hunk 6 above)
Comment (robfletcher) [inline orca-spring-batch/src/main/java/com/netflix/spinnaker/orca/batch/SpringBatchExecutionRunner.java]: Meant to take that out, my bad

**Pull request merged**